test/test_resolve.py

```diff
@@ -0,0 +1,209 @@
+import unittest
+
+from resolve_stream import QualifiedParameter, get_parameter, fully_resolve_parameter, lookup_parameter
+
+
+class TestResolveParameter(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        """
+        TODO - configure machine to machine interface in lieu of asset management resolution of streams and
+        reference designators
+        :return:
+        """
+        # config = yaml.load(open('../m2m_config.yml'))
+        # cls.m2m = MachineToMachine(config['url'], config['apiname'], config['apikey'])
+        # cls.stream_map = cls.m2m.streams()
+
+    def lookup(self, base_qp, *parameter_names):
+        """
+        Generate a qualified parameter from a parameter name and a base qualified parameter.
+        :param base_qp: QualifiedParameter template (missing Parameter field).
+        :param parameter_names: Parameter name(s) to lookup for given stream.
+        :return: fully qualified parameter
+        """
+        qp_list = []
+        for p in parameter_names:
+            qp_list.append(lookup_parameter(p, base_qp.stream, base_qp.refdes, base_qp.method))
+        if len(qp_list) == 1:
+            return qp_list[0]
+        return qp_list
+
+    def assertParameterResolved(self, parameter, expected_resolved, expected_unresolved):
+        """Assert parameter was resolved as expected."""
+        resolved, unresolved = fully_resolve_parameter(parameter)
+
+        # print '----- resolved -----'
+        # print '\n'.join(str(x) for x in sorted(resolved))
+        # print '----- expected -----'
+        # print '\n'.join(str(x) for x in sorted(expected_resolved))
+
+        self.assertListEqual(sorted(expected_resolved), sorted(resolved))
+        self.assertListEqual(sorted(expected_unresolved), sorted(unresolved))
+        self.assertNoDuplicates(resolved)
+        self.assertCorrectOrder(resolved)
+
+    def assertNoDuplicates(self, parameters):
+        """Assert there are no duplicate parameters listed."""
+        clean = set(parameters)
+        self.assertEqual(len(clean), len(parameters), 'parameter list has duplicates')
+
+    def assertCorrectOrder(self, parameters):
+        """Assert each parameter is listed before one that requires it."""
+        # starting with the last element, make sure that none of the previous parameters are required for resolution
+        parameters = parameters[::-1]
+        for i, p in enumerate(parameters):
+            resolved, unresolved = fully_resolve_parameter(p)
+            for r in resolved:
+                self.assertNotIn(p, parameters[i+1:], '%r should be listed after %r' % (p, r))
```

Comment:
should be iterating on `r` not `p`

Reply:
i usually try not to iterate on p as well.